shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -164,52 +164,114 @@ class JSBackend {
-      case TypeDef(Als, Primitive(alias), _, Primitive(name)) =>
-        S(alias -> name)
-      case _ => N
-    })
-    resolvedAliases.addAll(
-      Ls.from(relations flatMap { case (s, t) => s :: t :: Nil }).distinct map {
-        name =>
-          val visited = HashSet.empty[Str]
-          var current = name
-          while (!visited.contains(current)) {
-            visited += current
-            relations.get(current) match {
-              case N =>
-              case S(alias) =>
-                current = alias
+  // This function evaluates something like `T[A, B]`.
+  private def applyTypeAlias(name: Str, targs: Ls[Type]): Type =
+    typeAliasMap get name match {
+      case S(tparams -> body) =>
+        if (targs.length === tparams.length) {
+          val subs = collection.immutable.HashMap(
+            tparams zip targs map { case (k, v) => k -> v }: _*
+          )
+          substitute(body, subs)
+        } else {
+          throw new Error(
+            s"expect ${tparams.length} arguments but provided ${targs.length}"
+          )
+        }
+      case N => throw new Error(s"type $name is not defined")
+    }
+
+  // This function evaluates a type, removing all `AppliedType`s.
+  private def substitute(
+      body: Type,
+      subs: collection.immutable.HashMap[Str, Type] =
+        collection.immutable.HashMap.empty
+  ): Type = {
+    println(s"substitute $body")
+    body match {
+      case Applied(lhs, rhs) =>
+        val subsLhs = substitute(lhs, subs)
+        val subsRhs = substitute(rhs, subs)
+        if (subsLhs === lhs && subsRhs === rhs) {
+          body
+        } else {
+          Applied(subsLhs, subsRhs)
+        }
+      // Question: substitute arguments or the result?
+      case AppliedType(Primitive(name), targs) =>
+        applyTypeAlias(name, targs map { substitute(_, subs) })
+      case Function(lhs, rhs) =>
+        Function(substitute(lhs, subs), substitute(rhs, subs))
+      case Inter(lhs, rhs) =>
+        Inter(substitute(lhs, subs), substitute(rhs, subs))
+      case Record(fields) =>
+        Record(fields map { case (k, v) => k -> substitute(v, subs) })
+      case Union(lhs, rhs) =>
+        Union(substitute(lhs, subs), substitute(rhs, subs))
+      case Tuple(fields) =>
+        Tuple(fields map { case (k, v) => k -> substitute(v, subs) })
+      case Primitive(name) =>
+        subs get name match {
+          case N =>
+            typeAliasMap get name match {
+              case N              => Primitive(name)
+              case S(Nil -> body) => substitute(body, subs)
+              case S(tparams -> _) =>
+                throw new Error(
+                  s"type $name expects ${tparams.length} type parameters but nothing provided"
+                )
             }
-          }
-          name -> current
-      }
-    )
+          case S(ty) => ty
+        }
+      // For `Bot`, `Literal`, and `Top`, we don't need to substitute.
+      case _ => body
```

Comment:
See the comment I made to Tony a few days ago here: https://github.com/hkust-taco/mlscript/commit/e49c62e958224b9b1532a30fd45ac0909086fcbd#r60500415